config/locals.go

```diff
@@ -0,0 +1,313 @@
+package config
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/hcl2/hcl"
+	"github.com/hashicorp/hcl2/hcl/hclsyntax"
+	"github.com/hashicorp/hcl2/hclparse"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/gocty"
+
+	"github.com/gruntwork-io/terragrunt/errors"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/util"
+)
+
+// A consistent detail message for all "not a valid identifier" diagnostics. This is exactly the same as that returned
+// by terraform.
+const badIdentifierDetail = "A name must start with a letter and may contain only letters, digits, underscores, and dashes."
+
+// Local represents a single local name binding. This holds the unevaluated expression, extracted from the parsed file
+// (but before decoding) so that we can look for references to other locals before evaluating.
+type Local struct {
+	Name string
+	Expr hcl.Expression
+}
+
+// evaluateLocalsBlock is a routine to evaluate the locals block in a way to allow references to other locals. This
+// will:
+// - Extract a reference to the locals block from the parsed file
+// - Continuously evaluate the block until all references are evaluated, defering evaluation of anything that references
+//   other locals until those references are evaluated.
+// This returns a map of the local names to the evaluated expressions (represented as `cty.Value` objects). This will
+// error if there are remaining unevaluated locals after all references that can be evaluated has been evaluated.
+func evaluateLocalsBlock(
+	terragruntOptions *options.TerragruntOptions,
+	parser *hclparse.Parser,
+	hclFile *hcl.File,
+	filename string,
+) (map[string]cty.Value, error) {
+	diagsWriter := util.GetDiagnosticsWriter(parser)
+
+	localsBlock, diags := getLocalsBlock(hclFile)
+	if diags.HasErrors() {
+		diagsWriter.WriteDiagnostics(diags)
+		return nil, errors.WithStackTrace(diags)
+	}
+	if localsBlock == nil {
+		// No locals block referenced in the file
+		terragruntOptions.Logger.Printf("Did not find any locals block: skipping evaluation.")
+		return nil, nil
+	}
+
+	terragruntOptions.Logger.Printf("Found locals block: evaluating the expressions.")
+
+	locals, diags := decodeLocalsBlock(localsBlock)
+	if diags.HasErrors() {
+		terragruntOptions.Logger.Printf("Encountered error while decoding locals block into name expression pairs.")
+		diagsWriter.WriteDiagnostics(diags)
+		return nil, errors.WithStackTrace(diags)
+	}
+
+	// Continuously attempt to evaluate the locals until there are no more locals to evaluate, or we can't evaluate
+	// further.
+	evaluatedLocals := map[string]cty.Value{}
+	evaluated := true
+	for len(locals) > 0 && evaluated {
```

Comment:
Should we add a simple counter and ensure it stays under, say, 1000? Could be a nice safety valve in case of a bug in the code that leads to an infinite loop.

Actually, on that note, what if the code has:

```hcl
locals {
  a = local.b
  b = local.a
}
```

Reply:
This will fail because neither can be evaluated in the first loop, so `evaluated` will be `false`.

Reply:
> Should we add a simple counter and ensure it stays under, say, 1000? Could be a nice safety valve in case of a bug in the code that leads to an infinite loop.

Implemented.